modules/internal/ArrayViewRankChange.chpl

```diff
@@ -22,9 +22,337 @@
+
+    // the higher-dimensional domain that we're equivalent to
+    var downdomPid:int;
+    var downdomInst: downdomtype(downrank, idxType, stridable);
+
+    //
+    // TODO: If we put this expression into the variable declaration
+    // above, we get a memory leak.  File a future against this?
+    //
+    proc downdomtype(param rank: int, type idxType, param stridable: bool) type {
+      var a = dist.downdist.newRectangularDom(rank=rank, idxType=idxType,
+                                              stridable=stridable);
+      return a.type;
+    }
+
+    proc downrank param {
+      return collapsedDim.size;
+    }
+
+    inline proc downdom {
+      if _isPrivatized(downdomInst) then
+        return chpl_getPrivatizedCopy(downdomInst.type, downdomPid);
+      else
+        return downdomInst;
+    }
+
+    proc dsiBuildArray(type eltType) {
+      pragma "no auto destroy"
+      const downarr = _newArray(downdom.dsiBuildArray(eltType));
+      return new ArrayViewRankChangeArr(eltType  =eltType,
+      // TODO: Update once we start privatizing vvv
+                                        _DomPid = nullPid,
+                                        dom = this,
+                                        _ArrPid=downarr._pid,
+                                        _ArrInstance=downarr._instance,
+                                        collapsedDim=collapsedDim,
+                                        idx=idx,
+                                        ownsArrInstance=true);
+    }
+
+    // TODO: Use delegation feature for these?
+    // TODO: Can't all these be implemented in ChapelArray given dsiDim()?
+
+    proc dsiNumIndices {
+      return updom.dsiNumIndices;
+    }
+
+    proc dsiLow {
+      return updom.dsiLow;
+    }
+
+    proc dsiHigh {
+      return updom.dsiHigh;
+    }
+
+    proc dsiAlignedLow {
+      return updom.dsiAlignedLow;
+    }
+
+    proc dsiAlignedHigh {
+      return updom.dsiAlignedHigh;
+    }
+
+    proc dsiStride {
+      return updom.dsiStride;
+    }
+
+    proc dsiAlignment {
+      return updom.dsiAlignment;
+    }
+
+    proc dsiFirst {
+      return updom.dsiFirst;
+    }
+
+    proc dsiLast {
+      return updom.dsiLast;
+    }
+
+    proc dsiDim(upDim: int) {
+      var downDim = 0;
+      for d in 1..downrank {
+        if !collapsedDim(d) {
+          downDim += 1;
+          if (upDim == downDim) then
+            return downdom.dsiDim(d);
+        }
+      }
+      halt("Fell out of loop in dsiDim():", (upDim, downDim));
+    }
+
+    proc dsiDims() {
+      var upDims: rank*range(idxType, stridable=stridable);
+      var upDim = 1;
+      for param downDim in 1..downrank {
+        if !collapsedDim(downDim) {
+          upDims(upDim) = downdom.dsiDim(downDim);
+          upDim += 1;
+        }
+      }
+      return upDims;
+    }
+
+    proc dsiGetIndices() {
+      return updom.dsiGetIndices();
+    }
+
+    proc dsiSetIndices(inds) {
```

Comment:
You're correct that we haven't traditionally distinguished between the call that initializes a domain's indices and the one that re-assigns it -- both use dsiSetIndices() today.  I think that distinguishing those cases would be an interesting change to make (and one that we could make today), but don't think I should make it on this PR (even if time weren't a factor).  There's also a little bit of a question about what we would do differently in the two routines.

Note that I don't think dsiSetIndices() could be used to change the domain of a rank-change view array created by slicing an array because the only way that domain could be fetched is by applying a .domain to the array which returns a const ref to the domain.  But dsiSetIndices() could still be used for either initialization or assignment scenarios on a rank-change domain slice as follows:

````chapel
var D = {1..n, 1..n} dmapped Block(...);  // create a 2D domain
var A: [D] real;  // create an array over it
var X = A[3, ..];  // create X, a 1D view of 2D A

// create a new domain that has the same distribution as the rank-change view X.
// Note that this requires asking D's 2D block distribution to create a new 2D domain
// and for us to create a rank-change 1D view of it (because D's domain  map only knows
// how to create 2D things):
var D2 = {3..5} dmapped X.domain.dist;   // this will use dsiSetIndices() to establish D's indices
D2 = {9..11};  // D2 can be reassigned which will also call dsiSetIndices()
````